circuit_tutorials/plonky2/merkle_tree/src/main.rs

```diff
@@ -0,0 +1,238 @@
+}
+
+#[tokio::main]
+async fn main() {
+    // Uploads the circuit code to Sindri
+    compile_circuit().await;
+
+    // Uploads an input to the circuit consisting of a vector of 1024 leaves and an index value
+    // Proof artifacts are saved locally in a /data/ directory
+    let input_path: &str = "input_1024.json";
+    prove_circuit(input_path).await;
+
+    // Verifies the proof
+    let proof_path: &str = "./data/prove_out.json";
+    verify_proof(proof_path);
+}
+
+async fn compile_circuit() {
+    dotenv().expect("Failed to read .env file");
+    let api_key: String = std::env::var("SINDRI_API_KEY").unwrap();
+    let api_url_prefix: &str = "https://sindri.app/api/";
+    let api_version: &str = "v1/";
+    let api_url: String = api_url_prefix.to_owned() + api_version;
+
+    let mut contents = Vec::new();
+    {
+        // has to be scoped so that contents can be accessed after written to
+        let buffer = std::io::Cursor::new(&mut contents);
+        let enc = GzEncoder::new(buffer, Compression::default());
+        let mut tar = tar::Builder::new(enc);
+        tar.append_dir_all("merkle_tree", "./circuit")
+            .unwrap();
+    }
+
+    let part = Part::bytes(contents).file_name("filename.filetype");
+    let upload = reqwest::multipart::Form::new().part("files", part);
+
+    // Create new circuit
+    println!("Compiling circuit");
+    let client = Client::new();
+    let response = client
+        .post(format!("{api_url}circuit/create"))
+        .headers(headers_json(&api_key))
+        .multipart(upload)
+        .send()
+        .await
+        .unwrap();
+    assert_eq!(
+        &response.status().as_u16(),
+        &201u16,
+        "Expected status code 201"
+    );
+    let response_body = response.json::<Value>().await.unwrap();
+    let circuit_id = response_body["circuit_id"].as_str().unwrap();
+    println!("Circuit ID: {:?}", &circuit_id);
+
+    // Poll circuit detail until it has a status of Ready or Failed
+    let circuit_data = poll_status(
+        format!("circuit/{circuit_id}/detail"),
+        &api_url,
+        &api_key,
+        600,
+    )
+    .await;
+    if circuit_data["status"].as_str().unwrap().contains("Failed") {
+        println!("Circuit compilation failed.");
+        std::process::exit(1);
+    }
+
+    println!("Saving circuit details locally");
+    std::fs::create_dir_all("./data").unwrap();
+    let file = File::create("./data/compile_out.json").unwrap();
+    let mut writer = BufWriter::new(file);
+    serde_json::to_writer_pretty(&mut writer, &circuit_data).unwrap();
+    writer.flush().unwrap();
+}
+
+async fn prove_circuit(json_input_path: &str) {
+    {
+        dotenv().expect("Failed to read .env file");
+        let api_key: String = std::env::var("SINDRI_API_KEY").unwrap();
+        let api_url_prefix: &str = "https://sindri.app/api/";
+
+        let api_version: &str = "v1/";
+        let api_url: String = api_url_prefix.to_owned() + api_version;
```

Comment:
Perhaps move this common code which creates `api_url`, duplicated in several functions, into `main`? DRY

Reply:
done [a8c419d](https://github.com/Sindri-Labs/sindri-resources/pull/98/commits/a8c419d050fe294fcaeadf3d27ff3470a19f60a1)